File: Module_1/performance/print_primes.go
// A simple Go program to test CPU performance by
// generating a series of prime numbers

package main

import (
	"fmt"
	"math"
	"sync"
	"time"
	"os"
)

const (
	NUM_SERIES  = 100_000_000
	NUM_THREADS = 4
)

// isPrime checks if a number is a prime number.
func isPrime(number int) bool {
    limit := int(math.Sqrt(float64(number))) + 1
    for d := 2; d < limit; d++ {
        if number % d == 0 {
            return false
        }
    }
    return true
}
// printPrimes prints prime numbers in the range [start, stop).
func printPrimes(start, stop int, wg *sync.WaitGroup, threadName string) {
	defer wg.Done()
	for i := start; i < stop; i++ {
		if isPrime(i) {
			fmt.Printf("%s: %d\n", threadName, i)
		}
	}
}

func main() {
	var wg sync.WaitGroup
	chunkSize := NUM_SERIES / NUM_THREADS

	startTime := time.Now()

	for i := 0; i < NUM_THREADS; i++ {
		start := i * chunkSize
		stop := start + chunkSize
		threadName := fmt.Sprintf("Thread-%d", i+1)

		fmt.Fprintf(os.Stderr, "Created %s with start=%d, stop=%d\n", threadName, start, stop)

		wg.Add(1)
		go printPrimes(start, stop, &wg, threadName)
	}

	wg.Wait()
	duration := time.Since(startTime)
	fmt.Fprintf(os.Stderr, "%d Threads took %.6f seconds to complete\n", NUM_THREADS, duration.Seconds())
}

File: Module_1/performance/PrintPrimes.java
// A simple Java program to test CPU performance by
// generating a series of prime numbers

class PrimeTask implements Runnable {
    private long start;
    private long stop;

    public PrimeTask(long start, long stop) {
        this.start = start;
        this.stop = stop;
    }

    public boolean isPrime(long number) {
        var limit = (long)Math.sqrt(number) + 1;

        for (var d=2; d<limit; d++) {
            if (number % d == 0)
                return false;
        }
        return true;
    }

    @Override
    public void run() {
        var threadName = Thread.currentThread().getName();
        for (var i=start; i<stop; i++) {
            if (isPrime(i)) {
                System.out.printf("%s: %d\n", threadName, i);
            }
        }
    }
}

public class PrintPrimes {

    private static final int NUM_SERIES = 100_000_000;
    private static final int NUM_THREADS = 4;

    public static void main(String[] args) {
        Thread[] threads = new Thread[NUM_THREADS];
        var chunkSize = NUM_SERIES / NUM_THREADS;

        var startTime = System.currentTimeMillis();
        for (var i=0; i<NUM_THREADS; i++) {
            var start = i * chunkSize;
            var stop = start + chunkSize;
            var task = new PrimeTask(start, stop);

            threads[i] = new Thread(task);
            threads[i].start();
            System.err.printf("Created %s with start=%d, stop=%d\n", threads[i].getName(), start, stop);
        }

        for (Thread thread: threads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.err.println("Thread interrupted");
            }
        }
        var duration = (System.currentTimeMillis() - startTime) / 1000.0;
        System.err.printf("%d Threads took %.6f seconds to complete\n", NUM_THREADS, duration);
    }
}